File: project/android/apps/MnnLlmApp/app/src/main/java/com/alibaba/mls/api/download/DownloadPersistentData.java
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.

package com.alibaba.mls.api.download;

import android.content.Context;
import android.content.SharedPreferences;

import com.alibaba.mls.api.HfFileMetadata;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

public class DownloadPersistentData {

    public static final String METADATA_KEY = "meta_data";

    public static final String SIZE_TOTAL_KEY = "size_total";

    public static final String SIZE_SAVED_KEY = "size_saved";


    public static void saveMetaData(Context context, String modelId, List<HfFileMetadata> metaDataList) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        Gson gson = new Gson();
        String json = gson.toJson(metaDataList);
        editor.putString(METADATA_KEY, json);
        editor.apply();
    }

    public static List<HfFileMetadata> getMetaData(Context context, String modelId) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        String json = sharedPreferences.getString(METADATA_KEY, null);
        if (json != null) {
            Gson gson = new Gson();
            Type type = new TypeToken<List<HfFileMetadata>>() {}.getType();
            return gson.fromJson(json, type);
        }
        return null;
    }

    public static void saveDownloadSizeTotal(Context context, String modelId, long total) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putLong(SIZE_TOTAL_KEY, total);
        editor.apply();
    }

    public static long getDownloadSizeTotal(Context context, String modelId) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        return sharedPreferences.getLong(SIZE_TOTAL_KEY, 0);
    }

    public static void saveDownloadSizeSaved(Context context, String modelId, long saved) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putLong(SIZE_SAVED_KEY, saved);
        editor.apply();
    }

    public static long getDownloadSizeSaved(Context context, String modelId) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        return sharedPreferences.getLong(SIZE_SAVED_KEY, 0);
    }

    public static void removeProgress(Context context, String modelId) {
        modelId = HfFileUtils.getLastFileName(modelId);
        SharedPreferences sharedPreferences = context.getSharedPreferences("DOWNLOAD_" + modelId, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove(SIZE_SAVED_KEY);
        editor.remove(SIZE_TOTAL_KEY);
        editor.apply();
    }
}

File: apps/Android/MnnTaoAvatar/app/src/main/java/com/alibaba/mls/api/download/DownloadPersistentData.kt
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
package com.alibaba.mls.api.download

import android.content.Context
import com.alibaba.mls.api.HfFileMetadata
import com.alibaba.mls.api.download.DownloadFileUtils.getLastFileName
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

object DownloadPersistentData {
    const val METADATA_KEY: String = "meta_data"

    const val SIZE_TOTAL_KEY: String = "size_total"

    const val SIZE_SAVED_KEY: String = "size_saved"


    fun saveMetaData(context: Context, modelId: String?, metaDataList: List<HfFileMetadata?>?) {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()
        val gson = Gson()
        val json = gson.toJson(metaDataList)
        editor.putString(METADATA_KEY, json)
        editor.apply()
    }

    @JvmStatic
    fun getMetaData(context: Context, modelId: String?): List<HfFileMetadata>? {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        val json = sharedPreferences.getString(METADATA_KEY, null)
        if (json != null) {
            val gson = Gson()
            val type = object : TypeToken<List<HfFileMetadata?>?>() {}.type
            return gson.fromJson(json, type)
        }
        return null
    }

    @JvmStatic
    fun saveDownloadSizeTotal(context: Context, modelId: String?, total: Long) {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()
        editor.putLong(SIZE_TOTAL_KEY, total)
        editor.apply()
    }

    @JvmStatic
    fun getDownloadSizeTotal(context: Context, modelId: String?): Long {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        return sharedPreferences.getLong(SIZE_TOTAL_KEY, 0)
    }

    @JvmStatic
    fun saveDownloadSizeSaved(context: Context, modelId: String?, saved: Long) {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()
        editor.putLong(SIZE_SAVED_KEY, saved)
        editor.apply()
    }

    @JvmStatic
    fun getDownloadSizeSaved(context: Context, modelId: String?): Long {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        return sharedPreferences.getLong(SIZE_SAVED_KEY, 0)
    }

    @JvmStatic
    fun removeProgress(context: Context, modelId: String?) {
        var modelId = modelId
        modelId = getLastFileName(modelId)
        val sharedPreferences =
            context.getSharedPreferences("DOWNLOAD_$modelId", Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()
        editor.remove(SIZE_SAVED_KEY)
        editor.remove(SIZE_TOTAL_KEY)
        editor.apply()
    }
}
